### PR TITLE
ADM-730:[backend]fix: fix the logic to get last assignee and get historical assignees

### PR DESCRIPTION
**Pull request by mikeyangyun**
## Summary

...

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: backend/src/main/java/heartbeat/service/board/jira/JiraService.java [hunk 1]
```diff
@@ -494,31 +495,55 @@ private List<JiraCardDTO> getRealDoneCards(StoryPointsAndCycleTimeRequest reques
...
 		return assigneeSet;
 	}
 
-	private List<String> getHistoricalAssignees(URI baseUrl, String cardKey, String token) {
-		CardHistoryResponseDTO jiraCardHistory = jiraFeignClient.getJiraCardHistory(baseUrl, cardKey, token);
-		return jiraCardHistory.getItems().stream().map(item -> item.getActor().getDisplayName()).distinct().toList();
+	private String getLastHistoricalAssignee(CardHistoryResponseDTO jiraCardHistory) {
+		return jiraCardHistory.getItems()
+			.stream()
+			.filter(item -> AssigneeFilterMethod.ASSIGNEE_FIELD_ID.getDescription().equalsIgnoreCase(item.getFieldId()))
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/board/jira/JiraService.java:527]: Duplicated code with line 536

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java [hunk 2]
```diff
@@ -494,38 +494,80 @@ public static StoryPointsAndCycleTimeRequest.StoryPointsAndCycleTimeRequestBuild
...
 								.status(new Status(CardStepsEnum.DONE.getValue()))
 								.build())));
 	}
 
-	public static CardHistoryResponseDTO.CardHistoryResponseDTOBuilder CARD1_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER_METHOD() {
+	public static CardHistoryResponseDTO.CardHistoryResponseDTOBuilder CARD1_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER() {
+		return CardHistoryResponseDTO.builder()
+			.items(List.of(
+					new HistoryDetail(1673556350000L, "status", new Status(TESTING), new Status(REVIEW),
+							new HistoryDetail.Actor("da pei")),
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:522]: - what's the meaning of `"da pei"`
- duplicated code `"da pei"`, extract it to a fixture with meaningful varible name

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java [hunk 3]
```diff
@@ -494,38 +494,80 @@ public static StoryPointsAndCycleTimeRequest.StoryPointsAndCycleTimeRequestBuild
...
+					new HistoryDetail(1673556350001L, "assignee", new Status("song"), new Status("yun"),
+							new HistoryDetail.Actor("da pei"))));
+	}
+
+	public static CardHistoryResponseDTO.CardHistoryResponseDTOBuilder CARD2_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER() {
 		return CardHistoryResponseDTO.builder()
 			.items(List.of(
 					new HistoryDetail(1673556350000L, "status", new Status(TESTING), new Status(REVIEW),
 							new HistoryDetail.Actor("da pei")),
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:535]: btw, refactor it

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java [hunk 4]
```diff
@@ -494,38 +494,80 @@ public static StoryPointsAndCycleTimeRequest.StoryPointsAndCycleTimeRequestBuild
...
+					new HistoryDetail(1673556350001L, "assignee", new Status("kun"), new Status("yun"),
 							new HistoryDetail.Actor("da pei"))));
 	}
 
-	public static CardHistoryResponseDTO.CardHistoryResponseDTOBuilder CARD2_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER_METHOD() {
+	public static CardHistoryResponseDTO.CardHistoryResponseDTOBuilder CARD3_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER() {
 		return CardHistoryResponseDTO.builder()
 			.items(List.of(
 					new HistoryDetail(1673556350000L, "status", new Status(TESTING), new Status(REVIEW),
 							new HistoryDetail.Actor("da pei")),
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:548]: refactor

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java [hunk 5]
```diff
@@ -494,38 +494,80 @@ public static StoryPointsAndCycleTimeRequest.StoryPointsAndCycleTimeRequestBuild
...
+							new HistoryDetail.Actor("da pei"))));
+	}
+
+	public static AllDoneCardsResponseDTO.AllDoneCardsResponseDTOBuilder ALL_DONE_CARDS_RESPONSE_FOR_MULTIPLE_STATUS() {
+		return AllDoneCardsResponseDTO.builder()
+			.total("2")
+			.issues(List.of(
+					new JiraCard("ADM-475",
+							JiraCardField.builder()
+								.assignee(new Assignee("da pei"))
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:563]: refactor

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java [hunk 6]
```diff
@@ -494,38 +494,80 @@ public static StoryPointsAndCycleTimeRequest.StoryPointsAndCycleTimeRequestBuild
...
+								.status(new Status(CardStepsEnum.DONE.getValue()))
+								.build()),
+					new JiraCard("ADM-524",
+							JiraCardField.builder()
+								.assignee(new Assignee("xiao pei"))
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:568]: reactor

**Review by guzhongren — COMMENTED** [review]
refactor with the comments

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java (hunk 2 above)
Comment (mikeyangyun) [reply backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java]: "da pei" is just a name represents the assignee or the Actor of jira card changes.
I have extracted this to a constant in our fixture.

File: backend/src/main/java/heartbeat/service/board/jira/JiraService.java [hunk 7]
```diff
@@ -494,31 +497,46 @@ private List<JiraCardDTO> getRealDoneCards(StoryPointsAndCycleTimeRequest reques
...
+			else {
+				List<String> historicalAssignees = getHistoricalAssignees(jiraCardHistory);
+				assigneeSet.add(getLastHistoricalAssignee(historicalAssignees));
+			}
 		}
 
-		if (assignee != null && filterMethod.equals(AssigneeFilterMethod.HISTORICAL_ASSIGNEE.getDescription())) {
-			List<String> historyDisplayName = getHistoricalAssignees(baseUrl, doneCard.getKey(), request.getToken());
-			assigneeSet.addAll(historyDisplayName);
+		if (!useLastAssignee(assigneeFilter)) {
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/service/board/jira/JiraService.java:515]: put this part under `else`

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java [hunk 8]
```diff
@@ -64,6 +64,10 @@ public class JiraBoardConfigDTOFixture {
 
 	public static final String ASSIGNEE_NAME = "Zhang San";
 
+	public static final String DISPLAY_NAME_ONE = "da pei";
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:67]: what is `da pei` and `xiao pei`

Comment (guzhongren) [reply backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:67]: public static final String DISPLAY_NAME_ONE = "da pei"; ->public static final String DISPLAY_NAME_ONE = "Display name";

.....2

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java [hunk 9]
```diff
@@ -494,50 +498,92 @@ public static StoryPointsAndCycleTimeRequest.StoryPointsAndCycleTimeRequestBuild
...
+	public static CardHistoryResponseDTO.CardHistoryResponseDTOBuilder CARD3_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER() {
 		return CardHistoryResponseDTO.builder()
 			.items(List.of(
 					new HistoryDetail(1673556350000L, "status", new Status(TESTING), new Status(REVIEW),
-							new HistoryDetail.Actor("da pei")),
+							new HistoryDetail.Actor(DISPLAY_NAME_ONE)),
 					new HistoryDetail(1674556350000L, "status", new Status(DONE), new Status(TESTING),
-							new HistoryDetail.Actor("xiao pei"))));
+							new HistoryDetail.Actor(DISPLAY_NAME_ONE)),
+					new HistoryDetail(1673556350000L, "assignee", new Status("yun"), new Status(null),
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:555]: refactor `1673556350000L` with a readable variable

File: backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java [hunk 10]
```diff
@@ -816,32 +819,39 @@ void shouldGetRealDoneCardWhenCallGetStoryPointsAndCycleTimeWhenUseHistoricalAss
 		when(urlGenerator.getUri(any())).thenReturn(baseUrl);
 		when(jiraFeignClient.getJiraCards(any(), any(), anyInt(), anyInt(), any(), any())).thenReturn(allDoneCards);
 		when(jiraFeignClient.getJiraCardHistory(baseUrl, "ADM-475", token))
-			.thenReturn(CARD1_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER_METHOD().build());
+			.thenReturn(CARD1_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER().build());
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java:822]: given a `new line` between `given when then`

File: backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java [hunk 11]
```diff
@@ -851,15 +861,22 @@ void shouldGetRealDoneCardWhenCallGetStoryPointsAndCycleTimeWhenUseLastAssigneeF
...
 		when(jiraFeignClient.getJiraCardHistory(baseUrl, "ADM-524", token))
-			.thenReturn(CARD2_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER_METHOD().build());
+			.thenReturn(CARD2_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER().build());
+		when(jiraFeignClient.getJiraCardHistory(baseUrl, "ADM-520", token))
+			.thenReturn(CARD3_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER().build());
 		when(jiraFeignClient.getTargetField(baseUrl, "PLL", token)).thenReturn(ALL_FIELD_RESPONSE_BUILDER().build());
 
 		// then
-		CardCollection cardCollection = jiraService.getStoryPointsAndCycleTimeForDoneCards(request,
+		CardCollection cardCollection1 = jiraService.getStoryPointsAndCycleTimeForDoneCards(request,
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java:872]: 你们一起学一哈given when then 吧，实在对这个问题comment不动了

File: backend/src/main/java/heartbeat/service/board/jira/JiraService.java (hunk 7 above)
Comment (mikeyangyun) [reply backend/src/main/java/heartbeat/service/board/jira/JiraService.java]: Updated.

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java (hunk 8 above)
Comment (mikeyangyun) [reply backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java]: chenge the name to common pattern

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java (hunk 9 above)
Comment (mikeyangyun) [reply backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java]: Updated.

File: backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java (hunk 10 above)
Comment (mikeyangyun) [reply backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java:818]: Already update this pattern.

File: backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java (hunk 11 above)
Comment (mikeyangyun) [reply backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java:867]: 好，已经改过来啦

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java [hunk 12]
```diff
@@ -64,6 +64,20 @@ public class JiraBoardConfigDTOFixture {
 
 	public static final String ASSIGNEE_NAME = "Zhang San";
 
+	public static final String DISPLAY_NAME_ONE = "Da Pei";
+
+	public static final String DISPLAY_NAME_TWO = "Xiao Pei";
+
+	public static final String START_TIME = "1672556350000";
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:71]: 用一个可读性好的产量开搞，一串数字不知道代表什么时候

File: backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java [hunk 13]
```diff
@@ -584,19 +587,17 @@ void shouldGetCardsWhenCallGetStoryPointsAndCycleTimeWhenDoneTimeGreaterThanSele
...
 			.thenReturn(CARD_HISTORY_MULTI_RESPONSE_BUILDER().build());
 		when(jiraFeignClient.getJiraCardHistory(baseUrl, "2", token))
 			.thenReturn(CARD_HISTORY_DONE_TIME_GREATER_THAN_END_TIME_BUILDER().build());
 		when(jiraFeignClient.getTargetField(baseUrl, "PLL", token)).thenReturn(ALL_FIELD_RESPONSE_BUILDER().build());
-		// then
-
+		// when
 		CardCollection cardCollection = jiraService.getStoryPointsAndCycleTimeForDoneCards(
 				storyPointsAndCycleTimeRequest, jiraBoardSetting.getBoardColumns(), List.of("Zhang San"), "");
+		// then
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java:600]: 这些注释其实没必要，我们用空行表示

File: backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java [hunk 14]
```diff
@@ -793,73 +794,86 @@ public void shouldReturnJiraBoardConfigDTOWhenCallGetJiraBoardConfig() {
...
 		when(jiraFeignClient.getJiraCardHistory(baseUrl, "ADM-524", token))
-			.thenReturn(CARD2_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER_METHOD().build());
+			.thenReturn(CARD2_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER().build());
+		when(jiraFeignClient.getJiraCardHistory(baseUrl, "ADM-520", token))
+			.thenReturn(CARD3_HISTORY_FOR_HISTORICAL_ASSIGNEE_FILTER().build());
 		when(jiraFeignClient.getTargetField(baseUrl, "PLL", token)).thenReturn(ALL_FIELD_RESPONSE_BUILDER().build());
 
+		// when
+		CardCollection cardCollection1 = jiraService.getStoryPointsAndCycleTimeForDoneCards(request,
+				jiraBoardSetting.getBoardColumns(), List.of("Da Pei"), assigneeFilter);
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java:827]: Da Pei?

File: backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java [hunk 15]
```diff
@@ -967,9 +977,11 @@ void shouldGetRealDoneCardsGivenHistoryWithNoStatusFieldWhenCallGetStoryPointsAn
 			.thenReturn(CARD_HISTORY_WITH_NO_STATUS_FIELD().build());
 		when(jiraFeignClient.getTargetField(baseUrl, "PLL", token)).thenReturn(ALL_FIELD_RESPONSE_BUILDER().build());
 
-		// then
+		// when
 		CardCollection cardCollection = jiraService.getStoryPointsAndCycleTimeForDoneCards(request,
-				jiraBoardSetting.getBoardColumns(), List.of("da pei"), assigneeFilter);
+				jiraBoardSetting.getBoardColumns(), List.of("Da Pei"), assigneeFilter);
```
Comment (guzhongren) [inline backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java:982]: ???

**Review by guzhongren — APPROVED** [review]
改一哈，resolve一下就行，我后面再看

File: backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java (hunk 13 above)
Comment (mikeyangyun) [reply backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java:600]: 嗯嗯那我删除了吧

File: backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java (hunk 14 above)
Comment (mikeyangyun) [reply backend/src/test/java/heartbeat/service/jira/JiraServiceTest.java:827]: 我以为
image
这个是把人名规范化，还是说 public 改成 private ？

File: backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java (hunk 12 above)
Comment (mikeyangyun) [reply backend/src/test/java/heartbeat/service/jira/JiraBoardConfigDTOFixture.java:71]: 这块是一个比较常见的问题，下一张卡我会把测试中timestamp都用成date，增强可读性，暂时这张卡先fix issue

**Review by mjx20045912 — APPROVED** [review]
LGTM

**Pull request merged**